text/0034-ternary-hash.md

```diff
@@ -0,0 +1,472 @@
+## Sponges
+
+`CurlP` and `Kerl` are cryptographic sponge constructions. They are equipped with a memory state and a function that
+replaces the state memory using some input string. A portion of the memory state is then the output. In the sponge
+metaphor, the process of replacing the memory state by an input string is said to `absorb` the input, while the process
+of producing an output is said to `squeeze` out an output.
+
+The hash functions are expected to be used like this:
+
+- `curlp`
+  ```rust
+  // Create a CurlP instance with 81 rounds.
+  // This is equivalent to calling `CurlP::new(CurlPRounds::Rounds81)`.
+  let mut curlp = CurlP81::new();
+
+  // Assume there are some transaction trits, all zeroes for the sake of this example.
+  let transaction = TritBuf::<T1B1Buf>::zeros(6561);
+  let mut hash = TritBuf::<T1B1Buf>::zeros(243);
+
+  // Absorb the transaction.
+  curlp.absorb(&transaction);
+  // Squeeze out a hash.
+  curlp.squeeze_into(&mut hash);
+  ```
+
+- `kerl`
+  ```rust
+  // Create a Kerl instance.
+  let mut kerl = Kerl::new();
+
+  // `Kerl::digest` is a function that combines `Kerl::absorb` and `Kerl::squeeze`.
+  // `Kerl::digest_into` combines `Kerl::absorb` with `Kerl::squeeze_into`.
+  let hash = kerl.digest(&transaction);
+  ```
+
+The main proposal of this RFC are the `Sponge` trait and the `CurlP` and `Kerl` types that are implementing it.
+This RFC relies on the presence of the types `TritBuf` and `Trits`, as defined by
+[RFC36](https://github.com/iotaledger/bee-rfcs/blob/master/text/0036-ternary.md), which are assumed to be owning and
+borrowing collections of binary-encoded ternary in the `T1B1` encoding (one trit per byte).
+
+```rust
+/// The common interface of cryptographic hash functions that follow the sponge construction and that act on ternary.
+trait Sponge {
+    /// An error indicating that a failure has occured during a sponge operation.
+    type Error;
+
+    /// Absorb `input` into the sponge.
+    fn absorb(&mut self, input: &Trits) -> Result<(), Self::Error>;
+
+    /// Reset the inner state of the sponge.
+    fn reset(&mut self);
+
+    /// Squeeze the sponge into a buffer
+    fn squeeze_into(&mut self, buf: &mut Trits) -> Result<(), Self::Error>;
+
+    /// Convenience function using `Sponge::squeeze_into` to return an owned version of the hash.
+    fn squeeze(&mut self) -> Result<TritBuf, Self::Error> {
+        let mut output = TritBuf::zeros(HASH_LENGTH);
+        self.squeeze_into(&mut output)?;
+        Ok(output)
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge into a buffer, and reset the sponge in one go.
+    fn digest_into(&mut self, input: &Trits, buf: &mut Trits) -> Result<(), Self::Error> {
+        self.absorb(input)?;
+        self.squeeze_into(buf)?;
+        self.reset();
+        Ok(())
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge, and reset the sponge in one go.
+    /// Returns an owned version of the hash.
```

Comment:
"Returns an owned version of the hash" (the actual return type is `TritBuf` though, so maybe this requires a little explanation like: "Returns an owned version of the hash stored in a TritBuf". Wdyt?

Reply:
What would be the returned type if it wasn't `TritBuf` though ?

Reply:
well, since this RFC introduces `Hash`, and also introduces hash functions, that produce hashes, I would expect the return type to be `Hash`, and not `TritBuf`.

Reply:
Ahh, but it can't be `Hash`, the output of the sponge is of variable size.

Reply:
right

Reply:
but theoretically `digest` could return `Hash`.

Reply:
Yes. Up until recently there was still this `OUT_LEN` that would have totally prevented doing that. Even if this is technically possible to return an actual Hash, I wonder if we should do it, the API would look weird with different types idk.

Reply:
don't know either. definitly not an `approval` stopper, but we should maybe keep this in mind if we want to change that. I'ld personally like to have `Hash` returned for `digest` as it's convenient (I think), logical, and gives the introduction of `Hash` a usecase already.

Reply:
I tried to change but given the current definition of `Hash` this would mean a lot of new copies, since this is a very hot path I don't think that's wise to change :(

Reply:
alrighty :+1: Thanks for trying!!!